File: AULA_07/C/Dieta.c
#include <stdio.h>
#include <stdlib.h>

int main(){

  int A[3][7];
  double mediaDia[7], mediaRefeicao[3];
  FILE * f;
  int n,m;
  int i,j;
  int maior, maiorDia;

  n=3;
  m=7;

  f = fopen("calorias.csv","r");

  for( i=0;i<n;i++ )
    for( j=0;j<m;j++ )
      fscanf(f,"%d;",&A[i][j]);
  fclose(f);

  for( i=0;i<n;i++ ) mediaRefeicao[i]=0.0;
  for( i=0;i<m;i++ ) mediaDia[i]=0.0;

  maior = A[0][0];
  maiorDia = 0;
  for( i=0;i<n;i++ ){
    for( j=0;j<m;j++ ){
      mediaRefeicao[i]+=(double)A[i][j]/m;
      mediaDia[j]+=(double)A[i][j]/n;
      if( A[i][j] > maior ){
        maior = A[i][j];
        maiorDia = j;
      }
    }
  }

  printf("\tSeg.\tTer.\tQua.\tQui.\tSex.\tSab.\tDom.\tMedia\n");
  printf("Manhã\t");
  for( i=0;i<m;i++ ) printf("%d\t",A[0][i]);
  printf("%lf\n", mediaRefeicao[0]);
  printf("Tarde\t");
  for( i=0;i<m;i++ ) printf("%d\t",A[1][i]);
  printf("%lf\n", mediaRefeicao[1]);
  printf("Noite\t");
  for( i=0;i<m;i++ ) printf("%d\t",A[2][i]);
  printf("%lf\n", mediaRefeicao[2]);
  printf("Média\t");
  for( i=0;i<m;i++ ) printf("%.2lf\t",mediaDia[i]);
  printf("\n");

  printf("Dia de maior refeição: %d\n",maiorDia+1);

  for( i=0;i<m;i++ )
    if( A[0][i] > mediaRefeicao[0] || A[1][i] > mediaRefeicao[1] || A[2][i] > mediaRefeicao[2] )
      printf("No dia %d, comeu acima da média\n",i+1);

  return 0;
}

File: AULA_07/Haskell/Dieta.hs
import Data.List.Split
import Data.List
import System.IO

transposta :: [[a]] -> [[a]]
transposta ([]:_) = []
transposta x = (map head x) : transposta (map tail x)

maior [x,y,z]  
     | (x>=y) && (x>=z) = x
     | (y>=x) && (y>=z) = y
     | otherwise = z

argmax t = do
           let (x,y) = foldl (\ (x,i) (y,j) -> if x>y then (x,i) else (y,j) ) (-1,-1) (zip t [0..6])
           y+1

maiorCaloria t = argmax $ map maior t

acimaMedia t medias = map ( filter (\(x,y) -> y>x) ) (map (zip medias) t)

main = do
        linhas1 <- readFile "calorias.csv"
        let matrizStr x = map (splitOn ";") (lines x)
            calorias = map (map read) (matrizStr linhas1) :: [[Float]]
            mediaPeriodo = map (foldl (\acc x -> acc + x/7.0) 0) calorias
            mediaDia = map (foldl (\acc x -> acc + x/3.0) 0) (transposta calorias)
            tabela = zipWith (\x y -> x ++ [y]) calorias mediaPeriodo
            tabelaStr = zipWith (:) ["Manhã","Tarde","Noite"] (map (map show) tabela)
            acima = acimaMedia (transposta tabela) mediaPeriodo
        putStrLn "\tSeg.\tTer.\tQua.\tQui.\tSex.\tSab.\tDom.\tMedia"
        putStrLn $ (unlines (map (intercalate "\t") tabelaStr))
        putStrLn $ "\t" ++ (intercalate "\t" (map show mediaDia))
        putStrLn $ "Dia de maior caloria: " ++ (show (maiorCaloria (transposta tabela)))
        --putStrLn $ show (acimaMedia (transposta tabela) mediaPeriodo)
        let texto = filter (\(x,y) -> not (null x)) (zip (map (\t -> if null t then "" else " acima da média" ) acima) [1..7])
        putStrLn $ unlines (map (\(x,y) -> "Dia " ++ (show y) ++ x) texto)
